File: backend/app/schemas/terrain.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional, List, Dict, Any, Literal
from uuid import UUID


class CrossSection(BaseModel):
    station: float
    shape: List[List[float]]
    area: Optional[float] = None


class TerrainFeatures(BaseModel):
    centerline: Optional[Dict[str, Any]] = None
    cross_sections: Optional[List[CrossSection]] = None
    elevation_range: Optional[List[float]] = None
    slope_analysis: Optional[Dict[str, Any]] = None
    waterfront_line: Optional[Dict[str, Any]] = None
    demolition_boundary: Optional[Dict[str, Any]] = None
    farmland_boundary: Optional[Dict[str, Any]] = None


class TerrainStatus(BaseModel):
    status: Literal["pending", "processing", "completed", "failed"]
    progress: int = Field(ge=0, le=100)
    features: Optional[TerrainFeatures] = None


class TerrainResponse(BaseModel):
    id: UUID
    project_id: UUID
    file_type: str
    status: str
    features: Optional[TerrainFeatures] = None
    bounds: Optional[Dict[str, Any]] = None
    feature_count: Optional[int] = None
    warning: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)


class TerrainUploadResponse(BaseModel):
    id: UUID
    project_id: UUID
    file_type: str
    status: str
    features: Optional[Dict[str, Any]] = None
    bounds: Optional[Dict[str, Any]] = None
    feature_count: Optional[int] = None
    warning: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/terrain.ts
import client from './client'

export interface TerrainFeatures {
  centerline: unknown | null
  cross_sections: unknown[] | null
  elevation_range: [number, number] | null
  slope_analysis: Record<string, unknown> | null
}

export interface TerrainData {
  id: string
  project_id: string
  file_type: string
  status: string
  features: TerrainFeatures | null
  feature_count: number | null
}

export interface TerrainUploadResponse extends TerrainData {
  warning: { code: string; message: string } | null
}

export interface TerrainStatus {
  status: 'pending' | 'processing' | 'completed' | 'failed'
  progress: number
}

export const terrainApi = {
  upload: (projectId: string, file: File, onProgress?: (percent: number) => void) => {
    const formData = new FormData()
    formData.append('file', file)
    return client.post<TerrainUploadResponse>(`/projects/${projectId}/terrain`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
      onUploadProgress: (e) => {
        if (e.total && onProgress) {
          onProgress(Math.round((e.loaded * 100) / e.total))
        }
      },
    })
  },

  get: (projectId: string) =>
    client.get<TerrainData | null>(`/projects/${projectId}/terrain`),

  getStatus: (projectId: string) =>
    client.get<TerrainStatus>(`/projects/${projectId}/terrain/status`),
}
